pkg/events/event_ringbuffer.go

```diff
@@ -0,0 +1,173 @@
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package events
+
+import (
+	"time"
+
+	"github.com/apache/yunikorn-scheduler-interface/lib/go/si"
+)
+
+const latestUnset int64 = -1 << 63
+
+var now = time.Now
+
+// EventRingBuffer A specialized circular buffer to store event objects.
+//
+// Unlike to regular circular buffers, new entries can be added if the buffer is full. In this case,
+// the oldest entry is overwritten. This is not a classic enqueue operation, so it's named differently.
+//
+// Retrieving the records can be achieved with GetLatestEntriesCount and GetLatestEntries. Since these do not
+// remove the elements, they are not regular dequeue operations either.
+//
+// Entries have a maximum lifespan defined in nanoseconds. Cleanup of expired objects occurs when a call to
+// RemoveExpiredEntries is made.
+type EventRingBuffer struct {
+	events        []*si.EventRecord
+	capacity      int
+	noElements    int
+	head          int
+	tail          int
+	latest        int64
+	lifetimeNanos int64
+}
+
+// Add adds an event to the ring buffer. If the buffer is full, the oldest element is overwritten.
+// This method never fails.
+func (e *EventRingBuffer) Add(event *si.EventRecord) {
+	full := false
+	if e.noElements == e.capacity {
+		full = true
+	}
+	e.events[e.tail] = event
+	e.tail = e.next(e.tail)
+	e.latest = event.TimestampNano
+
+	if full {
+		e.head = e.tail
+		return
+	}
+
+	e.noElements++
+}
+
+// GetLatestEntriesCount returns most recent items. The amount is defined by "count".
+func (e *EventRingBuffer) GetLatestEntriesCount(count int) []*si.EventRecord {
+	if e.noElements == 0 {
+		return nil
+	}
+
+	records := make([]*si.EventRecord, 0)
+	for i := e.prev(e.tail); ; {
+		records = append(records, e.events[i])
+		if len(records) == count || i == e.head {
+			reverse(records)
+			return records
+		}
+
+		i = e.prev(i)
+	}
+}
+
+// GetLatestEntries returns the most recent items whose age is younger than the current time minus interval.
+func (e *EventRingBuffer) GetLatestEntries(interval time.Duration) []*si.EventRecord {
+	unixNow := now().UnixNano()
+
+	if e.noElements == 0 || unixNow-int64(interval) > e.latest {
+		return nil
+	}
+
+	startTime := unixNow - int64(interval)
+	records := make([]*si.EventRecord, 0)
+
+	for i := e.head; ; {
+		if e.events[i].TimestampNano >= startTime {
+			records = append(records, e.events[i])
+		}
+
+		next := e.next(i)
+		if next == e.tail {
+			break
+		}
+		i = next
+	}
+
+	return records
+}
+
+// RemoveExpiredEntries discards expired entries from the buffer. Current age of an object depends on the
+// TimestampNano field.
+// This method is expected to be called in regular intervals, which usually runs on a separate goroutine.
```

Comment:
From this comment, it looks like we will use `EventRingBuffer` in more than one gorutines. Do we want to add a lock when we update elements?

Reply:
We'll see how this type will be integrated with the existing code. Right now it's only a basic implementation. If it's accessed through an existing type which has locks, then it's not necessary to add it here.